File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = -1
    n = len(dungeons)
    visited = [False] * n
    
    def explore(visit, left):   
        nonlocal answer
        for i in range(n):
            if not visited[i] and left >= dungeons[i][0]:
                visited[i] = True
                visit.append(i)
                explore(visit, left-dungeons[i][1])
                visit.pop()
                visited[i] = False
        
        if answer < len(visit):
            answer = len(visit)
    
    explore([], k)
        
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    ArrayList<int[]> route = new ArrayList<>();
    
    void find_route(ArrayList<Integer> current,
                    int n,
                    boolean[] visited) {

        if (current.size() == n) {

            int[] temp = new int[n];
            for (int i = 0; i < n; i++) {
                temp[i] = current.get(i);
            }

            route.add(temp);
            return;
        }

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {

                visited[i] = true;
                current.add(i);

                find_route(current, n, visited);

                current.remove(current.size() - 1);
                visited[i] = false;
            }
        }
    }
    
    int find_dungeons(int[] route, int k, int[][] dungeons) {
        int cnt = 0;
        for(int i=0; i<route.length; i++) {
            if(dungeons[route[i]][0] <= k) {
                k -= dungeons[route[i]][1];
                cnt++;
            }
        }
        return cnt;
    }
    
    
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        int n = dungeons.length;
        boolean[] visited = new boolean[n];
        
        find_route(new ArrayList<>(), n, visited);
        ArrayList<Integer> result = new ArrayList<>();
        
        for(int i=0; i<route.size(); i++) {
            result.add(find_dungeons(route.get(i), k, dungeons));
        }
        
        
        return Collections.max(result);
    }
}